plugins/algoltek-aux/fu-algoltek-aux-device.c

```diff
@@ -0,0 +1,560 @@
+
+#include "fu-algoltek-aux-common.h"
+#include "fu-algoltek-aux-device.h"
+#include "fu-algoltek-aux-firmware.h"
+#include "fu-algoltek-aux-struct.h"
+
+struct _FuAlgoltekAuxDevice {
+	FuDpauxDevice parent_instance;
+	guint8 testoui;
+};
+
+G_DEFINE_TYPE(FuAlgoltekAuxDevice, fu_algoltek_aux_device, FU_TYPE_DPAUX_DEVICE)
+
+FuAlgoltekAuxDevice *
+fu_algoltek_aux_device_new(FuDpauxDevice *device)
+{
+	FuAlgoltekAuxDevice *self = g_object_new(FU_TYPE_ALGOLTEK_AUX_DEVICE, NULL);
+	if (device != NULL)
+		fu_device_incorporate(FU_DEVICE(self), FU_DEVICE(device));
+	return self;
+}
+
+static gboolean
+fu_algoltek_dpaux_device_write(FuAlgoltekAuxDevice *self, GByteArray *buf, guint delayms, GError **error)
+{
+	fu_device_sleep(FU_DEVICE(self), delayms);
+	if (!fu_dpaux_device_write(FU_DPAUX_DEVICE(self),
+				  0x80000,
+				  buf->data,
+				  buf->len,
+				  ALGOLTEK_DEVICE_AUX_TIMEOUT,
+				  error))
+		return FALSE;
+	
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_dpaux_device_read(FuAlgoltekAuxDevice *self, GByteArray *buf, GError **error)
+{
+	fu_device_sleep(FU_DEVICE(self), 20);
+	if (!fu_dpaux_device_read(FU_DPAUX_DEVICE(self),
+				  0x80010,
+				  buf->data,
+				  buf->len,
+				  ALGOLTEK_DEVICE_AUX_TIMEOUT,
+				  error))
+		return FALSE;
+	
+	return TRUE;
+}
+
+
+static guint16
+gen_crc16(guint16 BData, guint16 crc)
+{
+	guint16 BFlag = 0;
+
+	for (guint16 i= 0 ; i < 8; i++ ) {
+		BFlag = BData ^ (crc >> 8);
+		crc <<= 1;
+		
+		if (BFlag & 0x80) {
+			crc ^= FU_ALGOLTEK_AUX_CMD_CRC_POLINOM;
+			
+		}
+		BData <<= 1;
+		
+	}
+	
+	return crc;
+}
+
+static GByteArray *
+fu_algoltek_aux_device_rdv(FuAlgoltekAuxDevice *self, GError **error)
+{	
+	g_autoptr(GByteArray) replay = g_byte_array_new();
+	g_autoptr(GByteArray) version_data = g_byte_array_new();
+	guint copyCount = 0;
+
+	fu_byte_array_set_size(replay, 16, 0x0);
+	fu_byte_array_set_size(version_data, 64, 0x0);
+
+	g_autoptr(GByteArray) st = fu_struct_algoltek_aux_rdv_cmd_address_pkt_new();
+	gsize length = st->len - 3;
+
+	fu_struct_algoltek_aux_rdv_cmd_address_pkt_set_i2c_address(st, 0x51);
+	fu_struct_algoltek_aux_rdv_cmd_address_pkt_set_sublen(st, length);
+	fu_struct_algoltek_aux_rdv_cmd_address_pkt_set_address(st, 0);
+	fu_struct_algoltek_aux_rdv_cmd_address_pkt_set_len(st, length);
+	fu_struct_algoltek_aux_rdv_cmd_address_pkt_set_cmd(st, FU_ALGOLTEK_AUX_CMD_RDV);
+
+	for (guint i = 0; i < 4; i++) {
+	if(!fu_algoltek_dpaux_device_write(FU_DPAUX_DEVICE(self), st, 20, error)){
+		g_prefix_error(error, "\naux dpcd write failed:\n");
+		return "aux dpcd rdv failed:" ;
+	}
+
+	if(!fu_algoltek_dpaux_device_read(FU_DPAUX_DEVICE(self), replay, error)){
+		g_prefix_error(error, "\naux dpcd read failed:\n");
+		return "aux dpcd rdv failed:" ;
+	}
+
+	if (i == 0) {
+		memcpy(version_data->data + copyCount, replay->data + 2, 14);
+		copyCount += 14;	
+	} else {
+		memcpy(version_data->data + copyCount, replay->data + 0, 16);
+		copyCount += 16;
+	}
+	g_print(replay->data);
```

Comment:
do you mean g_debug? Can you give some context?

Reply:
Sorry it's for my debug, no need now.
Deleted it.